File: ALODWebUtility/LookUps/LookUp.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Web.UI.WebControls;
using System.Xml;
using AjaxControlToolkit;
using ALOD.Core.Domain.Common;
using ALOD.Core.Domain.Modules.Lod;
using ALOD.Data;

namespace ALODWebUtility.LookUps
{
    public class LookUp : LookUpTable
    {
        // Public Function GetMemberTypes() As LookUp
        //     Dim cmd As DbCommand = DataStore.GetSqlStringCommand("select  memberType, memberDescr from core_lkupMemberType Order By id")
        //     GetCollection(DataStore, "memberType", cmd)
        //     Return Me
        // End Function

        public IEnumerable<ICD9Code> GetCategory(int headingId)
        {
            ICD9CodeDao icdDao = new NHibernateDaoFactory().GetICD9CodeDao();
            var query = from r in icdDao.GetAll() where r.ParentId == headingId && r.Active select r;
            return query;
        }

        public IEnumerable<ICD9Code> GetChapter()
        {
            ICD9CodeDao icdDao = new NHibernateDaoFactory().GetICD9CodeDao();
            var query = from r in icdDao.GetAll() where r.Code == null && r.ParentId == null && r.Active orderby r.SortOrder select r;

            return query;
        }

        public LookUp GetGroupsByCompo(string compo)
        {
            DbCommand cmd = DataStore.GetStoredProcCommand("core_group_sp_GetByCompo");
            DataStore.AddInParameter(cmd, "@compo", DbType.String, compo);
            GetCollection(DataStore, "groups_compo_" + compo, cmd);
            return this;
        }

        public DataSets.LookupDataTable GetGroupsByCompoAsDataSet(string compo)
        {
            return GetGroupsByCompo(compo).ToDataSet();
        }

        public IEnumerable<ICD9Code> GetICDChildren(int parentId)
        {
            ICD9CodeDao icdDao = new NHibernateDaoFactory().GetICD9CodeDao();
            var query = from r in icdDao.GetAll() where r.ParentId == parentId && r.Active select r;
            return query;
        }

        public CascadingDropDownNameValue[] GetICDCodeCDDValues(StringDictionary categoryValues, string parentCategory)
        {
            if (!categoryValues.ContainsKey(parentCategory))
            {
                return null;
            }

            int selectedParentId = Convert.ToInt32(categoryValues[parentCategory]);
            IEnumerable<ICD9Code> children = GetICDChildren(selectedParentId);
            List<CascadingDropDownNameValue> values = new List<CascadingDropDownNameValue>();
            string text = string.Empty;

            foreach (ICD9Code item in children)
            {
                if (string.IsNullOrEmpty(item.Code))
                {
                    text = item.Description;
                }
                else
                {
                    text = item.Description + " - " + item.Code;
                }

                values.Add(new CascadingDropDownNameValue(text, item.Id.ToString()));
            }

            return values.ToArray();
        }

        public List<ListItem> GetICDIncident(int codeId)
        {
            if (codeId == 0)
            {
                return null;
            }

            ICD9CodeDao icdDao = new NHibernateDaoFactory().GetICD9CodeDao();
            ICD9Code code = icdDao.GetById(codeId);
            List<ListItem> values = new List<ListItem>();

            List<NatureOfIncident> incidentValues = icdDao.GetAssociatedNatureOfIncidentValues(codeId);

            if (incidentValues != null && incidentValues.Count > 0)
            {
                foreach (NatureOfIncident v in incidentValues)
                {
                    values.Add(new ListItem(v.Text, v.Value));
                }
            }
            else
            {
                values.Add(new ListItem("Illness", "Illness"));
                values.Add(new ListItem("Injury", "Injury"));
                values.Add(new ListItem("Injury MVA", "Injury-MVA"));
                values.Add(new ListItem("Disease", "Disease"));
                values.Add(new ListItem("Death", "Death"));
            }

            // Commented out legacy code preserved from VB version
            // If code.Id < 467 Then...

            return values;
        }

        public DataSet GetManagedGroups(int groupId)
        {
            return DataStore.ExecuteDataSet("core_group_sp_GetManaged", groupId);
        }

        public DataSet GetManagedGroupsDropDown(int groupId)
        {
            return DataStore.ExecuteDataSet("core_group_sp_GetManagedDropDown", groupId);
        }

        public DataSet GetStates()
        {
            DbCommand cmd = DataStore.GetSqlStringCommand("select state, state_name from core_lkupStates order by country desc, state_name");
            return DataStore.ExecuteDataSet(cmd);
        }

        public DataSet GetViewedGroups(int groupId)
        {
            return DataStore.ExecuteDataSet("core_group_sp_GetViewBy", groupId);
        }

        public LookUp GetWorkflowActionTypes()
        {
            DbCommand cmd = DataStore.GetSqlStringCommand("SELECT type, text FROM core_lkupWorkflowAction ORDER BY text");
            GetCollection(DataStore, "workflowActions", cmd);
            return this;
        }

        public DataSets.LookupDataTable GetWorkflowActionTypesAsDataSet()
        {
            return GetWorkflowActionTypes().ToDataSet();
        }

        public DataSet SearchUsers(int userId, string ssn, string name, byte status, int role, int unitId, bool showAllUsers)
        {
            return DataStore.ExecuteDataSet("core_user_sp_GetManagedUsers", userId, ssn, name, status, role, unitId, showAllUsers);
        }

        #region Caching

        public DataSets.LookupDataTable ToDataSet()
        {
            DataSets.LookupDataTable data = new DataSets.LookupDataTable();
            DataSets.LookupRow row;

            foreach (LookUpRow item in _rows)
            {
                row = data.NewLookupRow();
                row.key = item.text;
                row.value = item.value;
                data.Rows.Add(row);
            }

            return data;
        }

        public void ToXml(ref XmlNode parentNode)
        {
            XmlDocument doc = parentNode.OwnerDocument;

            string section = "";
            int currentId = 0;

            foreach (LookUpRow row in _rows)
            {
                XmlNode node = doc.CreateElement("Row");

                node.Attributes.Append(doc.CreateAttribute("Text"));
                node.Attributes["Text"].Value = row.text;

                node.Attributes.Append(doc.CreateAttribute("Value"));
                node.Attributes["Value"].Value = row.value;

                parentNode.AppendChild(node);
            }
        }

        protected void GetCollection(string key, string procName, params object[] parameterValues)
        {
            GetCollection(DataStore, key, DataStore.GetStoredProcCommand(procName, parameterValues));
        }

        protected void GetCollection(SqlDataStore adapter, string key, DbCommand cmd)
        {
            if (Cache[key] != null)
            {
                // we have this item in cache
                _rows = (List<LookUpRow>)Cache[key];
            }
            else
            {
                // get it from the db
                adapter.ExecuteReader(LookUpRowReader, cmd);
                Cache.Insert(key, _rows);
            }
        }

        #endregion
    }
}

File: ALODWebUtility/LookUps/LookUp.vb
Imports System.Collections.Specialized
Imports System.Data.Common
Imports System.Web.UI.WebControls
Imports System.Xml
Imports AjaxControlToolkit
Imports ALOD.Core.Domain.Common
Imports ALOD.Core.Domain.Modules.Lod
Imports ALOD.Data

Namespace LookUps

    Public Class LookUp
        Inherits LookUpTable

        'Public Function GetMemberTypes() As LookUp
        '    Dim cmd As DbCommand = DataStore.GetSqlStringCommand("select  memberType, memberDescr from core_lkupMemberType Order By id")
        '    GetCollection(DataStore, "memberType", cmd)
        '    Return Me
        'End Function

        Public Function GetCategory(ByVal headingId As Integer) As IEnumerable(Of ICD9Code)

            Dim icdDao As ICD9CodeDao = New NHibernateDaoFactory().GetICD9CodeDao()
            Dim query = From r In icdDao.GetAll() Where r.ParentId = headingId And r.Active Select r
            Return query

        End Function

        Public Function GetChapter() As IEnumerable(Of ICD9Code)

            Dim icdDao As ICD9CodeDao = New NHibernateDaoFactory().GetICD9CodeDao()
            Dim query = From r In icdDao.GetAll() Where r.Code Is Nothing And r.ParentId Is Nothing And r.Active Select r Order By r.SortOrder

            Return query

        End Function

        Public Function GetGroupsByCompo(ByVal compo As String) As LookUp
            Dim cmd As DbCommand = DataStore.GetStoredProcCommand("core_group_sp_GetByCompo")
            DataStore.AddInParameter(cmd, "@compo", DbType.String, compo)
            GetCollection(DataStore, "groups_compo_" + compo, cmd)
            Return Me
        End Function

        Public Function GetGroupsByCompoAsDataSet(ByVal compo As String) As DataSets.LookupDataTable
            Return GetGroupsByCompo(compo).ToDataSet()
        End Function

        Public Function GetICDChildren(ByVal parentId As Integer) As IEnumerable(Of ICD9Code)
            Dim icdDao As ICD9CodeDao = New NHibernateDaoFactory().GetICD9CodeDao()
            Dim query = From r In icdDao.GetAll() Where r.ParentId = parentId And r.Active Select r
            Return query
        End Function

        Public Function GetICDCodeCDDValues(ByVal categoryValues As StringDictionary, ByVal parentCategory As String) As CascadingDropDownNameValue()

            If (Not categoryValues.ContainsKey(parentCategory)) Then
                Return Nothing
            End If

            Dim selectedParentId As Integer = CType(categoryValues(parentCategory), Integer)
            Dim children As IEnumerable(Of ICD9Code) = GetICDChildren(selectedParentId)
            Dim values As New List(Of CascadingDropDownNameValue)
            Dim text As String = String.Empty

            For Each item As ICD9Code In children
                If (String.IsNullOrEmpty(item.Code)) Then
                    text = item.Description
                Else
                    text = item.Description & " - " & item.Code
                End If

                values.Add(New CascadingDropDownNameValue(text, item.Id))
            Next

            Return values.ToArray()

        End Function

        Public Function GetICDIncident(ByVal codeId As Integer) As List(Of ListItem)
            If (IsNothing(codeId) Or codeId = 0) Then
                Return Nothing
            End If

            Dim icdDao As ICD9CodeDao = New NHibernateDaoFactory().GetICD9CodeDao()
            Dim code As ICD9Code = icdDao.GetById(codeId)
            Dim values As New List(Of ListItem)

            Dim incidentValues As List(Of NatureOfIncident) = icdDao.GetAssociatedNatureOfIncidentValues(codeId)

            If (incidentValues IsNot Nothing AndAlso incidentValues.Count > 0) Then
                For Each v As NatureOfIncident In incidentValues
                    values.Add(New ListItem(v.Text, v.Value))
                Next
            Else
                values.Add(New ListItem("Illness", "Illness"))
                values.Add(New ListItem("Injury", "Injury"))
                values.Add(New ListItem("Injury MVA", "Injury-MVA"))
                values.Add(New ListItem("Disease", "Disease"))
                values.Add(New ListItem("Death", "Death"))
            End If

            'If code.Id < 467 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 466 And code.Id < 475 Then
            '    values.Add(New ListItem("Illness", "Illness"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 474 And code.Id < 485 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id = 485 Then
            '    values.Add(New ListItem("Illness", "Illness"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 485 And code.Id < 493 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id = 493 Then
            '    values.Add(New ListItem("Illness", "Illness"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 493 And code.Id < 780 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 779 And code.Id < 800 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Illness", "Illness"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 799 And code.Id < 812 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 811 And code.Id < 919 Then
            '    values.Add(New ListItem("Injury", "Injury"))
            '    values.Add(New ListItem("Injury MVA", "Injury-MVA"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 918 And code.Id < 925 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 924 And code.Id < 982 Then
            '    values.Add(New ListItem("Injury", "Injury"))
            '    values.Add(New ListItem("Injury MVA", "Injury-MVA"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 981 And code.Id < 1043 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf code.Id > 1042 And code.Id < 1045 Then
            '    values.Add(New ListItem("Injury", "Injury"))
            '    values.Add(New ListItem("Injury MVA", "Injury-MVA"))
            '    values.Add(New ListItem("Death", "Death"))
            'Else
            '    values.Add(New ListItem("Illness", "Illness"))
            '    values.Add(New ListItem("Injury", "Injury"))
            '    values.Add(New ListItem("Injury MVA", "Injury-MVA"))
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'End If

            'If codeId < 467 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 466 And codeId < 475 Then
            '    values.Add(New ListItem("Illness", "Illness"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 474 And codeId < 485 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId = 485 Then
            '    values.Add(New ListItem("Illness", "Illness"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 485 And codeId < 493 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId = 493 Then
            '    values.Add(New ListItem("Illness", "Illness"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 493 And codeId < 780 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 779 And codeId < 800 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Illness", "Illness"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 799 And codeId < 812 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 811 And codeId < 919 Then
            '    values.Add(New ListItem("Injury", "Injury"))
            '    values.Add(New ListItem("Injury MVA", "Injury-MVA"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 918 And codeId < 925 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 924 And codeId < 982 Then
            '    values.Add(New ListItem("Injury", "Injury"))
            '    values.Add(New ListItem("Injury MVA", "Injury-MVA"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 981 And codeId < 1043 Then
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'ElseIf codeId > 1042 And codeId < 1045 Then
            '    values.Add(New ListItem("Injury", "Injury"))
            '    values.Add(New ListItem("Injury MVA", "Injury-MVA"))
            '    values.Add(New ListItem("Death", "Death"))
            'Else
            '    values.Add(New ListItem("Illness", "Illness"))
            '    values.Add(New ListItem("Injury", "Injury"))
            '    values.Add(New ListItem("Injury MVA", "Injury-MVA"))
            '    values.Add(New ListItem("Disease", "Disease"))
            '    values.Add(New ListItem("Death", "Death"))
            'End If

            Return values

        End Function

        Public Function GetManagedGroups(ByVal groupId As Integer) As DataSet
            Return DataStore.ExecuteDataSet("core_group_sp_GetManaged", groupId)
        End Function

        Public Function GetManagedGroupsDropDown(ByVal groupId As Integer) As DataSet
            Return DataStore.ExecuteDataSet("core_group_sp_GetManagedDropDown", groupId)
        End Function

        Public Function GetStates() As DataSet
            Dim cmd As DbCommand = DataStore.GetSqlStringCommand("select state, state_name from core_lkupStates order by country desc, state_name")
            Return DataStore.ExecuteDataSet(cmd)
        End Function

        Public Function GetViewedGroups(ByVal groupId As Integer) As DataSet
            Return DataStore.ExecuteDataSet("core_group_sp_GetViewBy", groupId)
        End Function

        Public Function GetWorkflowActionTypes() As LookUp
            Dim cmd As DbCommand = DataStore.GetSqlStringCommand("SELECT type, text FROM core_lkupWorkflowAction ORDER BY text")
            GetCollection(DataStore, "workflowActions", cmd)
            Return Me
        End Function

        Public Function GetWorkflowActionTypesAsDataSet() As DataSets.LookupDataTable
            Return GetWorkflowActionTypes().ToDataSet()
        End Function

        Public Function SearchUsers(ByVal userId As Integer, ByVal ssn As String, ByVal name As String, ByVal status As Byte, ByVal role As Integer, ByVal unitId As Integer, ByVal showAllUsers As Boolean) As DataSet
            Return DataStore.ExecuteDataSet("core_user_sp_GetManagedUsers", userId, ssn, name, status, role, unitId, showAllUsers)
        End Function

#Region "Caching"

        Public Function ToDataSet() As DataSets.LookupDataTable

            Dim data As New DataSets.LookupDataTable
            Dim row As DataSets.LookupRow

            For Each item As LookUpRow In _rows
                row = data.NewLookupRow
                row.key = item.text
                row.value = item.value
                data.Rows.Add(row)
            Next

            Return data

        End Function

        Public Sub ToXml(ByRef parentNode As XmlNode)

            Dim doc As XmlDocument = parentNode.OwnerDocument

            Dim section As String = ""
            Dim currentId As Integer = 0

            For Each row As LookUpRow In _rows

                Dim node As XmlNode = doc.CreateElement("Row")

                node.Attributes.Append(doc.CreateAttribute("Text"))
                node.Attributes("Text").Value = row.text

                node.Attributes.Append(doc.CreateAttribute("Value"))
                node.Attributes("Value").Value = row.value

                parentNode.AppendChild(node)

            Next

        End Sub

        Protected Sub GetCollection(ByVal key As String, ByVal procName As String, ByVal ParamArray parameterValues As Object())
            GetCollection(DataStore, key, DataStore.GetStoredProcCommand(procName, parameterValues))
        End Sub

        Protected Sub GetCollection(ByVal adapter As SqlDataStore, ByVal key As String, ByVal cmd As DbCommand)

            If (Cache(key) IsNot Nothing) Then
                'we have this item in cache
                _rows = CType(Cache(key), List(Of LookUpRow))
            Else
                'get it from the db
                adapter.ExecuteReader(AddressOf LookUpRowReader, cmd)
                Cache.Insert(key, _rows)
            End If

        End Sub

#End Region

    End Class

End Namespace